---
title: "impute_weight_aggregate"
output: html_document
editor_options: 
  chunk_output_type: console
---

started: Alexey Larionov, 27Feb2017  
last updated: Alexey Larionov, 20Sep2019

# Summary  

Perform SKAT-style data pre-processing and aggregation  
Preprocessing steps done in the following order:  

- Read and check bc8 data  
- *No need in inverting genotypes* because there was no variants with AFs > 0.5  
- Impute missed genotypes (2 x total AF = mean genotype, as "fixed" imputation in SKAT)  
- Weight variants by total AFs (dbeta[AF,1,25], default weight in SKAT, *assumes AF < 0.5* !!!)  
- Aggregate preprocessed (imputed-weighted) variants in bc8 genes (just summation: as in "burden" SKAT style)  

##Input & output data:  
- 162 variants (62 genes) x 664 samples (467BC = 232UBC + 235CBC and 197NFE)  

# start_section

<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>

```{r set_width, echo=F}

#eval=F

options(width = 999)
# https://stackoverflow.com/questions/36845178/width-of-r-code-chunk-output-in-rmarkdown-files-knitr-ed-to-html

```

```{r start_section}

# Time stamp
Sys.time()

# Clenan-up
rm(list=ls())
graphics.off()

# Base folder
library(knitr)
base_folder="/Users/alexey/Documents/wecare/ampliseq/v04_ampliseq_nfe/s17_polr"
opts_knit$set(root.dir = base_folder)
#setwd(base_folder)

# Load libraries
library(dplyr)

```

# load_data

```{r load_data}

source_folder="/Users/alexey/Documents/wecare/ampliseq/v04_ampliseq_nfe/s16_crude_stats"
load(paste(source_folder, "s02_crude_genes_stats.RData", sep="/"))
base_folder="/Users/alexey/Documents/wecare/ampliseq/v04_ampliseq_nfe/s17_polr"
rm(source_folder)

```

# check_data

```{r check_data}

ls()

dim(genotypes.mx)
dim(variants.df)
dim(phenotypes.df)
dim(genes.df)

# Check consistency of rownames and colnames
sum(colnames(genotypes.mx) != rownames(phenotypes.df))
sum(rownames(genotypes.mx) != rownames(variants.df))

```

# impute_missed_data

Substitute NA with mean genotype per variant (= duplicated **total** AFs).  
Equivalent to the "fixed" imputation approach in SKAT (see impute.method parameter for SKAT function).  
Note: No AFs > 0.5 !!!

```{r impute_missed_data}

# Check AFs
max(variants.df$total_AF)

# Prepare matrix for output
genotypes_imp.mx <- genotypes.mx

# Fraction of missed data 
sum(is.na(genotypes.mx))
sum(is.na(genotypes.mx))/
  (nrow(genotypes.mx)*ncol(genotypes.mx)) # ~3.6%

# Impute missed data
for(var in 1:nrow(genotypes_imp.mx)){

    # Get AF
    af <- variants.df[var,"total_AF"]
    
    # Impute missed genotypes
    genotypes <- genotypes_imp.mx[var,]
    2*af -> genotypes[is.na(genotypes)]
    genotypes -> genotypes_imp.mx[var,]

    # Clean-up
    rm(af, genotypes)

}

# Check the result
sum(is.na(genotypes_imp.mx)) # 0
hist(genotypes_imp.mx, lab=T, ylim=c(0,1.3e5))
hist(genotypes_imp.mx[genotypes_imp.mx>0], lab=T, ylim=c(0,4.5e3))

# Clean-up
rm(var)

```

# weight_variants_by_af

Weight by **total** AF using dbeta(AF,1,25)  
Similar to the default weighting in SKAT  
(see parameter weights.beta=c[1,25] for SKAT function)  

```{r weight_variants_by_af}

# Prepare tables for output
variants.df <- data.frame(variants.df, weight=NA)
genotypes_imp_wt.mx <- genotypes_imp.mx

# Apply weighting
for(var in 1:nrow(genotypes_imp_wt.mx)){
  
  # Calculate weight
  weight <- dbeta(variants.df[var, "total_AF"], 1, 25)
  
  # Write weight to variants table
  weight -> variants.df[var,"weight"]
  
  # Apply weights to genotypes table
  weight * genotypes_imp.mx[var,] -> genotypes_imp_wt.mx[var,]
  
  # Clean-up
  rm(weight)

}

# Explore weights
sum(is.na(variants.df$weight))
mean(variants.df$weight)
hist(variants.df$weight, lab=T, ylim=c(0,160))

# Explore weighted genotypes
min(genotypes_imp_wt.mx)
max(genotypes_imp_wt.mx)

hist(genotypes_imp_wt.mx, lab=T, ylim=c(0,1.2e5))
hist(genotypes_imp_wt.mx[genotypes_imp_wt.mx>0], lab=T, ylim=c(0,4.5e3))

hist(genotypes_imp_wt.mx[genotypes_imp_wt.mx>0.5], lab=T, ylim=c(0,3e2))

# clean-up
rm(var, genotypes_imp.mx)

```

# aggregate_imputed_weighted_data_per_gene

```{r aggregate_imputed_weighted_data_per_gene}

# Get list of genes (there should be no NAs !)
genes <- unique(as.character(variants.df$SYMBOL))
length(genes)
sum(is.na(genes))

# Prepare results matrix
genes_data.mx <- matrix(nrow=0, ncol=ncol(genotypes_imp_wt.mx))
colnames(genes_data.mx) <- colnames(genotypes_imp_wt.mx)

for(gene in genes){
  
  #gene="ATM"
  
  # Get list of variants
  vars <- variants.df$SYMBOL == gene
  
  # Get genotypes matrix
  gt.mx <- genotypes_imp_wt.mx[vars, , drop=FALSE]

  # Aggregate in "burden" style (no NAs is expected after imputation)
  aggr_gene <- apply(gt.mx, 2, sum)
  
  # Add to results matrix
  genes_data.mx <- rbind(genes_data.mx, aggr_gene)
  
  # In-loop clean-up
  rm(vars, gt.mx, aggr_gene)
  
}

# Add rownames
genes -> rownames(genes_data.mx)

# Check result
dim(genes_data.mx)
genes_data.mx[1:5,1:5]
sum(is.na(genes_data.mx))

# Explore result
min(genes_data.mx)
max(genes_data.mx)

x <- as.vector(genes_data.mx)
x[order(x, decreasing = TRUE)][1:25]

# Samples with high weighted genotypes
for(ind in which(genes_data.mx > 40)){

  cat(genes_data.mx[ind]," - ")
  col=ceiling(ind / nrow(genes_data.mx))
  row=which(genes_data.mx[,col]>40)
  genes_data.mx[row,col]
  cat(rownames(genes_data.mx)[row]," - ")
  cat(colnames(genes_data.mx)[col],"\n")
  
  rm(col, row)
  
}

# Check SETX
hist(genes_data.mx[genes.df$gene=="SETX",], lab=T, ylim=c(0,7e2))
phenotypes.df[c("102_S466_L008","142_S501_L008"),
              c("cc")]

# 
hist(genes_data.mx, lab=T, ylim=c(0,4.5e4))
hist(genes_data.mx[genes_data.mx > 1 & genes_data.mx < 40],
     lab=T, ylim=c(0,3e2))

# Clean-up
rm(gene, genes, x, ind)

```

# data_summary

```{r data_summary}

# Rename genes.df
genes_info.df <- genes.df
rm(genes.df)

ls()

dim(genotypes.mx)
dim(genotypes_imp_wt.mx)
dim(genes_data.mx)
dim(genes_info.df)
dim(variants.df)
dim(phenotypes.df)

# Check consistency of rownames and colnames
sum(colnames(genotypes.mx) != rownames(phenotypes.df))
sum(colnames(genes_data.mx) != rownames(phenotypes.df))

sum(rownames(genes_info.df) != rownames(genes_data.mx))

sum(rownames(genotypes.mx) != rownames(genotypes_imp_wt.mx))
sum(rownames(genotypes.mx) != rownames(variants.df))

```

# save_data

```{r save_data}

save.image(paste(base_folder, "s01_impute_weight_aggregate.RData", sep="/"))

```

# final_section

```{r final_section}

ls()
sessionInfo()
Sys.time()

```
