---
title: "WES: crude counts per variant"
author: "Alexey Larionov"
date: "31 January 2021"
output:
  html_document:
    toc: true
    toc_float: true
editor_options:
  chunk_output_type: console
---

# Summary

- Calculate crude counts: totals, and separately for NFE, cases and controls  
- Trend test for the crude counts  
- FDR for all variants  

Check call rates (per variant and per sample) before and after filtering.  

<style>
pre{
  overflow-x: auto;
}
pre code{
  word-wrap: normal;
  white-space: pre;
}
</style>

```{r echo=F}
options(width=999)
```

# Start section

```{r}

# Time
Sys.time()

# Memory
gc()

# Clean up
rm(list=ls())
graphics.off()

# Options
options(stringsAsFactors = F)

# Folders
base_folder <- "/Users/alexey/Documents"
project_folder <- file.path(base_folder,"wecare","final_analysis_2021","reanalysis_wo_danish_2021","s02_wes_wecare_nfe")
scripts_folder <- file.path(project_folder,"scripts","s08_crude_counts")
setwd(scripts_folder)
data_folder <- file.path(project_folder,"data","s08_crude_counts")
# data_folder should already exist and contain a text file with BC8 genes

# Libraries and functions
library(dplyr)
library(reshape2)
library(ggplot2)
kj
source(file.path(project_folder,"scripts","f00_functions","f01_qqunif_plot.R"))
source(file.path(project_folder,"scripts","f00_functions","f02_tryCatchAdv.R"))

```

# Load data

```{r}

load(file.path(project_folder,"data","s07_relatedness_and_pca","s06_update_main_dataset.RData"))

# update folders
base_folder <- "/Users/alexey/Documents"
project_folder <- file.path(base_folder,"wecare","final_analysis_2021","reanalysis_wo_danish_2021","s02_wes_wecare_nfe")
scripts_folder <- file.path(project_folder,"scripts","s08_crude_counts")
data_folder <- file.path(project_folder,"data","s08_crude_counts")

```

# Check data

```{r}

ls()

# Check objects dimentions
dim(genotypes.mx)
dim(phenotypes.df)
dim(variants.df)

# Check that the tables and matrices are in sync
sum(phenotypes.df$wes_id != colnames(genotypes.mx))
sum(variants.df$ID != rownames(genotypes.mx))

```

# Prepare indices for cases and controls

```{r}

table(phenotypes.df$cc, useNA = "always")

nfe <- phenotypes.df$cc == -1
controls <- phenotypes.df$cc == 0
cases <- phenotypes.df$cc == 1

```

# Prepare matrix for results

```{r}

col_names <- c("id",
               "total_ac","total_an","total_af",
               "nfe_ac","nfe_an","nfe_af",
               "controls_ac","controls_an","controls_af",
               "cases_ac","cases_an","cases_af",
               "trend_direction","trend_test_p",
               "test_status","test_message")

variants_crude_counts.mx <- matrix(nrow=0,ncol=length(col_names))

colnames(variants_crude_counts.mx) <- col_names

rm(col_names)

```

# Calculate crude counts and Fisher test

takes ~1hr

```{r}

# For each variant
for(var in 1:nrow(variants.df)){
  
  # for testing
  #var <- 1
  
  # Get variant ID
  id <- variants.df[var,"ID"]
  
  # Get vector of genotypes
  genotypes <- genotypes.mx[var,]
  
  # Calculate crude counts
  total_ac <- sum(genotypes, na.rm=T)
  total_an <- 2*sum(!is.na(genotypes))
  total_af <- total_ac / total_an

  nfe_ac <- sum(genotypes[nfe], na.rm=T)
  nfe_an <- 2*sum(!is.na(genotypes[nfe]))
  nfe_af <- nfe_ac / nfe_an

  controls_ac <- sum(genotypes[controls], na.rm=T)
  controls_an <- 2*sum(!is.na(genotypes[controls]))
  controls_af <- controls_ac / controls_an

  cases_ac <- sum(genotypes[cases], na.rm=T)
  cases_an <- 2*sum(!is.na(genotypes[cases]))
  cases_af <- cases_ac / cases_an

  # Determine the trend direction
  "none" -> trend_direction
  if(nfe_af < controls_af & controls_af < cases_af) "risk" -> trend_direction
  if(nfe_af > controls_af & controls_af > cases_af) "protective" -> trend_direction
  if(nfe_af < cases_af & trend_direction=="none") "toward_risk" -> trend_direction
  if(nfe_af > cases_af & trend_direction=="none") "toward_protection" -> trend_direction
  
  # Proportions trend test: nfe-ubc-cbc
  ac_prop_trend <- c(nfe_ac,controls_ac,cases_ac)
  an_prop_trend <- c(nfe_an,controls_an,cases_an)
  
  try_trend_test <- tryCatchAdv(prop.trend.test(ac_prop_trend, an_prop_trend))
  
  trend_test_p <- try_trend_test$value$p.value
  test_status <- try_trend_test$status
  test_message <- try_trend_test$message
  
  # Compile result 
  result <- c(id,
     total_ac,total_an,total_af,
     nfe_ac,nfe_an,nfe_af,
     controls_ac,controls_an,controls_af,
     cases_ac,cases_an,cases_af,
     trend_direction,trend_test_p,test_status,test_message)
  
  # Add result to the table
  variants_crude_counts.mx <- rbind(variants_crude_counts.mx, result)

  # Clean-up (to avoid chances of counts carry-over between records)
  rm(id,genotypes,
     total_ac,total_an,total_af,
     nfe_ac,nfe_an,nfe_af,
     controls_ac,controls_an,controls_af,
     cases_ac,cases_an,cases_af,
     trend_direction,try_trend_test,
     ac_prop_trend, an_prop_trend,
     trend_test_p,test_status,test_message,
     result)
  
} # Next variant

# Check result
dim(variants_crude_counts.mx)
head(variants_crude_counts.mx)

# Clean-up
rm(var, nfe, cases, controls, tryCatchAdv)

```

# Update numeric fields

```{r}

# Make data frame
variants_crude_counts.df <- as.data.frame(variants_crude_counts.mx)
dim(variants_crude_counts.df)

# Update field types
variants_crude_counts.df$id <- as.character(variants_crude_counts.df$id)

variants_crude_counts.df$total_ac <- as.integer(variants_crude_counts.df$total_ac)
variants_crude_counts.df$total_an <- as.integer(variants_crude_counts.df$total_an)
variants_crude_counts.df$total_af <- as.numeric(variants_crude_counts.df$total_af)

variants_crude_counts.df$nfe_ac <- as.integer(variants_crude_counts.df$nfe_ac)
variants_crude_counts.df$nfe_an <- as.integer(variants_crude_counts.df$nfe_an)
variants_crude_counts.df$nfe_af <- as.numeric(variants_crude_counts.df$nfe_af)

variants_crude_counts.df$controls_ac <- as.integer(variants_crude_counts.df$controls_ac)
variants_crude_counts.df$controls_an <- as.integer(variants_crude_counts.df$controls_an)
variants_crude_counts.df$controls_af <- as.numeric(variants_crude_counts.df$controls_af)

variants_crude_counts.df$cases_ac <- as.integer(variants_crude_counts.df$cases_ac)
variants_crude_counts.df$cases_an <- as.integer(variants_crude_counts.df$cases_an)
variants_crude_counts.df$cases_af <- as.numeric(variants_crude_counts.df$cases_af)

variants_crude_counts.df$trend_direction <- as.character(variants_crude_counts.df$trend_direction)
variants_crude_counts.df$trend_test_p <- as.numeric(variants_crude_counts.df$trend_test_p)

variants_crude_counts.df$test_status <- as.character(variants_crude_counts.df$test_status)
variants_crude_counts.df$test_message <- as.character(variants_crude_counts.df$test_message)

str(variants_crude_counts.df)

rownames(variants_crude_counts.df) <- variants_crude_counts.df$id

# Clean-upy
rm(variants_crude_counts.mx)

```

# Explore AFs and trends

```{r}

data.df <-variants_crude_counts.df[,c("nfe_af","controls_af","cases_af")]
c("nfe","controls","cases") -> colnames(data.df)
dim(data.df)

molten.df <- melt(data.df)
dim(molten.df)

colnames(molten.df) <- c("group","af")
head(molten.df)

# Make plot (also can use position dodge/identity)
ggplot(molten.df, aes(x=af, fill=group)) + 
  geom_histogram(alpha=0.5, bins=15, position="dodge2", aes(y=..density..)) +
  labs(title="AF-s per group",
       x = "AF", y ="Density")

# Explore trend_direction
table(variants_crude_counts.df$trend_direction, useNA = "always")

# Clean-up
rm(data.df,molten.df)

```

# Calculate adjusted p-values 

```{r}

trend_test_p_adj <- p.adjust(variants_crude_counts.df$trend_test_p, method="BH")

variants_crude_counts.df <- cbind(variants_crude_counts.df, trend_test_p_adj)

rm(trend_test_p_adj)

```

# Add results to variants table

```{r}

sum(variants.df$ID != variants_crude_counts.df$id)

variants.df <- cbind(variants.df, variants_crude_counts.df)

variants.df <- variants.df %>% select(-id)

rm(variants_crude_counts.df)

```

# Look at the top variants

ToDo: *look at vep_SYMBOL=="."*

```{r}

min(variants.df$trend_test_p_adj)

variants.df %>% 
  select(ID,vep_SYMBOL,vep_gnomAD_NFE_AF,total_af,nfe_af,controls_af,cases_af,trend_direction,trend_test_p,trend_test_p_adj) %>% 
  arrange(trend_test_p_adj) %>% 
  filter(vep_gnomAD_NFE_AF <= 0.01 & total_af <=0.01 & trend_test_p <= 0.005 & trend_direction == "risk")

```

# QQ plot

```{r}

qqunif.plot(variants.df$trend_test_p)

rm(qqunif.plot)

```

# Check data

```{r}

ls()

# Check objects dimentions
dim(genotypes.mx)
dim(phenotypes.df)
dim(variants.df)

# Check that the tables and matrices are in sync
sum(phenotypes.df$wes_id != colnames(genotypes.mx))
sum(variants.df$ID != rownames(genotypes.mx))

```

# Save results

```{r}

save.image(file.path(data_folder,"s01_crude_counts_per_variant.RData"))

```

# Final section

```{r}

sessionInfo()
Sys.time()
gc()

```
